### ISSUE TITLE
Rule proposal: detect unnecessary use of fragments

### ISSUE THREAD

**Issue by moretti** [yannickcr/eslint-plugin-react#2042]
The idea behind fragments is to group a list of children without adding extra nodes to the DOM. 
Fragments are completely redundant when wrapping a single child, so the rule should detect examples like:

```.js
<>
  <div />
</>
```

```.js
<Fragment>
  <div />
</Fragment>
```

and flag them as a warning/error.

Example of valid usage of fragments:
```.js
<>
  <div />
  <div />
</>
```
(any case where `children.length > 1`)

**Comment by pawelnvk** [comment]
I would love to work on this one!

**Comment by golopot** [comment]
Other cases:

1. Looks useless 
```js
const Foo = () => (
  <div>
    <>
      <div />
      <div />
    </>
  </div>
)
```

2. Looks useless, but arguable.
```js
const Goo = () => (
  <div>
    <>
      <div />
      <div />
    </>
    <section />
  </div>
)
```
3. Not useless if it is a child of a non-dom component.
```js
const Hoo = () => (
  <Moo>
    <>
      <div />
      <div />
    </>
  </Moo>
)
```

**Comment by ljharb** [comment]
I do think it’s always useless to pass a fragment as a child to an html element; the rule should warn on that too.

**Comment by golopot** [comment]
Here comes the problem of whether a tag is a html element, also noted in https://github.com/yannickcr/eslint-plugin-react/issues/1752#issuecomment-377800024. Maybe in this case it is ok to maintain a list of html tags in this repo, since in #1752 an omission of an html tag name is false positive, whereas in this rule, an omission is false negative.

**Comment by ljharb** [comment]
It’s an html element if the tag name is lowercase and doesn’t have a dot in it; that’s how jsx works - that comment is about knowing whether it’s a valid tag name or not.

**Comment by ljharb** [comment]
I suppose the autofix for that case wouldn’t be as safe if it wasn’t a valid tag name; but we could still warn on it.

**Comment by TkDodo** [comment]
When can we expect this to be released? It's been some time since this has been finished ...

**Comment by paolostyle** [comment]
What if I have a component that looks like this (very trivial example, usually the stuff in the map is a bit more complicated):
```
const Component = ({ someArray }) => {
  return (
    <>
      {someArray.map(item => <span>{item}</span>)}
    </>
  );
```
The rule reports an error but I think there shouldn't be one. Obviously it is possible that there will be 0, 1 or more children inside fragment, but I think there should be way to disable the rule for these situations through config.

**Comment by ljharb** [comment]
@paolostyle in that case, you're already missing a `key` inside the array; however components in react 16 can render arrays directly, so there's no need for the fragment.

**Comment by paolostyle** [comment]
That was just an artificial example, I always have keys in arrays in actual code because I use a linting rule for that but, I don't think that's relevant here.

My whole codebase is in Typescript and I can't return an array from function components (and 95% of them are function components), there's a couple of links to Typescript and React types GH issues in this SO thread: https://stackoverflow.com/questions/46643517/return-react-16-array-elements-in-typescript. 

I suppose my example is actually just a workaround so I just won't use this rule until it's fixed in TS/React types. Looking at how old this issue is I probably won't ever use it but oh well.

**Comment by ljharb** [comment]
TypeScript having broken types for React is something that TS needs to fix; my suggested workaround would be to not use TS :-p (but that SO answer tells you how to patch the types so it works)
